### PR TITLE
Assign Goonj Chapter To Contact

### PR DESCRIPTION
**Pull request by belwalshubham**
Assign the Goonj Chapter to both the organization and the individual contact

## Summary by CodeRabbit

- **New Features**
	- Enhanced event handling for chapter group assignments.
	- Added methods for retrieving and assigning contact groups based on state ID.
	- Improved logic for adding contacts to groups.
	- Expanded data retrieval in the collection camp cron function.

- **Bug Fixes**
	- Updated logging for scenarios where no contact group is found.
	- Enhanced logging for missing IDs during group assignments.

- **Documentation**
	- Clarified method functionalities and logging details for better understanding.

### CODE REVIEW

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 1]
```diff
@@ -30,7 +32,10 @@ class InstitutionCollectionCampService extends AutoSubscriber {
   public static function getSubscribedEvents() {
     return [
       '&hook_civicrm_fieldOptions' => 'setIndianStateOptions',
-      '&hook_civicrm_pre' => 'generateInstitutionCollectionCampQr',
+      '&hook_civicrm_pre' => [
+        ['assignChapterGroupToIndividual'],
+      ],
+      // '&hook_civicrm_pre' => 'generateInstitutionCollectionCampQr',
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php]: Why we need this?

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 2]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
...
+   *
+   */
+  private static function addContactToGroup($contactId, $groupId) {
+    if ($contactId && $groupId) {
+      GroupContact::create(FALSE)
+        ->addValue('contact_id', $contactId)
+        ->addValue('group_id', $groupId)
+        ->addValue('status', 'Added')
+        ->execute();
+      \Civi::log()->info("Added contact_id: $contactId to group_id: $groupId");
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:108]: Dont think we need this

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 3]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
...
+        self::addContactToGroup($organizationId, $groupId);
+      }
+    }
+  }
+
+  /**
+   *
+   */
+  private static function addContactToGroup($contactId, $groupId) {
+    if ($contactId && $groupId) {
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:102]: Let's return it before

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 4]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
...
+      GroupContact::create(FALSE)
+        ->addValue('contact_id', $contactId)
+        ->addValue('group_id', $groupId)
+        ->addValue('status', 'Added')
+        ->execute();
+      \Civi::log()->info("Added contact_id: $contactId to group_id: $groupId");
+    }
+    else {
+      \Civi::log()->info("Failed to add contact to group. Invalid contact_id or group_id");
+    }
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:112]: Use try catch

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 5]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
...
+  /**
+   *
+   */
+  private static function getChapterGroupForState($stateId) {
+    $stateContactGroups = Group::get(FALSE)
+      ->addSelect('id')
+      ->addWhere('Chapter_Contact_Group.Use_Case', '=', 'chapter-contacts')
+      ->addWhere('Chapter_Contact_Group.Contact_Catchment', 'CONTAINS', $stateId)
+      ->execute();
+    $stateContactGroup = $stateContactGroups->first();
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:56]:
```suggestion
      ->execute()->first();
```

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 6]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
     ];
   }
 
+  /**
+   *
+   */
+  private static function getChapterGroupForState($stateId) {
+    $stateContactGroups = Group::get(FALSE)
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:51]:
```suggestion
    $stateContactGroup = Group::get(FALSE)
```

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 7]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
...
+      ->addSelect('id')
+      ->addWhere('Chapter_Contact_Group.Use_Case', '=', 'chapter-contacts')
+      ->addWhere('Chapter_Contact_Group.Contact_Catchment', 'CONTAINS', $stateId)
+      ->execute();
+    $stateContactGroup = $stateContactGroups->first();
+
+    if (!$stateContactGroup) {
+      \CRM_Core_Error::debug_log_message('No chapter contact group found for state ID: ' . $stateId);
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:59]: not needed

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 8]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
...
+
+      $fallbackGroups = Group::get(FALSE)
+        ->addWhere('Chapter_Contact_Group.Use_Case', '=', 'chapter-contacts')
+        ->addWhere('Chapter_Contact_Group.Fallback_Chapter', '=', 1)
+        ->execute();
+      $stateContactGroup = $fallbackGroups->first();
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:65]:
```suggestion
        ->execute()->first();
```

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 9]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
...
+      ->addSelect('id')
+      ->addWhere('Chapter_Contact_Group.Use_Case', '=', 'chapter-contacts')
+      ->addWhere('Chapter_Contact_Group.Contact_Catchment', 'CONTAINS', $stateId)
+      ->execute();
+    $stateContactGroup = $stateContactGroups->first();
+
+    if (!$stateContactGroup) {
+      \CRM_Core_Error::debug_log_message('No chapter contact group found for state ID: ' . $stateId);
+
+      $fallbackGroups = Group::get(FALSE)
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:61]:
```suggestion
      $fallbackGroup = Group::get(FALSE)
```

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php [hunk 10]
```diff
@@ -39,6 +44,74 @@ public static function getSubscribedEvents() {
...
+        ->addWhere('Chapter_Contact_Group.Use_Case', '=', 'chapter-contacts')
+        ->addWhere('Chapter_Contact_Group.Fallback_Chapter', '=', 1)
+        ->execute();
+      $stateContactGroup = $fallbackGroups->first();
+
+      \Civi::log()->info('Assigning fallback chapter contact group: ' . $stateContactGroup['title']);
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionCollectionCampService.php:67]: not needed

**Review by tarunnjoshi — CHANGES_REQUESTED** [review]

**Comment by belwalshubham** [comment]
Done With All Feedbacks, Merging

**Pull request merged**